File: python/pia/resources/options.py
"""Official PIA SDK - Derivatives & Options Analytics Resource."""

from __future__ import annotations

import urllib.parse
from typing import Any, Optional

from ..errors import ValidationError
from ..transport import AsyncTransport, SyncTransport
from ..types import OptionChainResponse, OptionGexResponse, OptionSummaryResponse


class OptionsResource:
    """Synchronous Options & Derivatives Analytics resource."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def get_chain(self, symbol: str) -> OptionChainResponse:
        """Fetches the real-time option chain for an underlying asset symbol."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        data = self._transport.request(
            "/api/v1/options/chain", method="GET", params={"symbol": clean}
        )
        return OptionChainResponse.from_dict(data)

    def get_gex(self, symbol: str) -> OptionGexResponse:
        """Fetches Gamma Exposure (GEX) profile and zero-gamma inflection levels."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        raw = self._transport.request(
            "/api/v1/options/gex", method="GET", params={"symbol": clean}
        )
        payload = raw.get("data") if isinstance(raw, dict) else raw
        if isinstance(raw, dict) and isinstance(payload, list):
            payload = {"symbol": clean, "items": payload}
        elif isinstance(payload, dict):
            payload = {**payload, "symbol": payload.get("symbol") or clean}
        elif isinstance(raw, dict):
            payload = {**raw, "symbol": raw.get("symbol") or clean}
        return OptionGexResponse.from_dict(payload if isinstance(payload, dict) else {"symbol": clean, "data": payload})

    def get_summary(self) -> OptionSummaryResponse:
        """Retrieves overall options market activity and put/call sentiment summary."""
        data = self._transport.request("/api/v1/options/summary", method="GET")
        return OptionSummaryResponse.from_dict(data)


class AsyncOptionsResource:
    """Asynchronous Options & Derivatives Analytics resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def get_chain(self, symbol: str) -> OptionChainResponse:
        """Asynchronously fetches real-time option chain for an underlying asset symbol."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        data = await self._transport.request(
            "/api/v1/options/chain", method="GET", params={"symbol": clean}
        )
        return OptionChainResponse.from_dict(data)

    async def get_gex(self, symbol: str) -> OptionGexResponse:
        """Asynchronously fetches Gamma Exposure (GEX) profile."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        raw = await self._transport.request(
            "/api/v1/options/gex", method="GET", params={"symbol": clean}
        )
        payload = raw.get("data") if isinstance(raw, dict) else raw
        if isinstance(raw, dict) and isinstance(payload, list):
            payload = {"symbol": clean, "items": payload}
        elif isinstance(payload, dict):
            payload = {**payload, "symbol": payload.get("symbol") or clean}
        elif isinstance(raw, dict):
            payload = {**raw, "symbol": raw.get("symbol") or clean}
        return OptionGexResponse.from_dict(payload if isinstance(payload, dict) else {"symbol": clean, "data": payload})

    async def get_summary(self) -> OptionSummaryResponse:
        """Asynchronously retrieves overall options market activity and put/call summary."""
        data = await self._transport.request("/api/v1/options/summary", method="GET")
        return OptionSummaryResponse.from_dict(data)

File: typescript/src/resources/options.ts
/**
 * Official PIA SDK - Derivatives & Options Analytics Resource
 */

import { normalizeSymbol } from "../symbols";
import type { HttpTransport } from "../http/transport";
import type {
  OptionChainResponse,
  OptionGexResponse,
  OptionSummaryResponse,
  RequestOptions,
} from "../types";

export class OptionsResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Fetches the real-time option chain for an underlying asset symbol.
   *
   * @param symbol Underlier ticker (e.g. "AAPL", "NVDA", "SPX")
   */
  public async getChain(
    symbol: string,
    options?: RequestOptions
  ): Promise<OptionChainResponse> {
    const clean = normalizeSymbol(symbol);
    const params = new URLSearchParams({ symbol: clean });
    const raw = await this.transport.request<any>(
      `/api/v1/options/chain?${params.toString()}`,
      "GET",
      undefined,
      options
    );
    const payload = raw?.data || raw;
    return {
      symbol: payload?.symbol || clean,
      underlying_price: payload?.underlying_price,
      expirations: payload?.expirations || [],
      contracts: payload?.contracts || payload?.items || [],
    };
  }

  /**
   * Fetches Gamma Exposure (GEX) profile and key zero-gamma levels.
   *
   * @param symbol Underlier ticker (e.g. "SPX", "NVDA", "QQQ")
   */
  public async getGex(
    symbol: string,
    options?: RequestOptions
  ): Promise<OptionGexResponse> {
    const clean = normalizeSymbol(symbol);
    const params = new URLSearchParams({ symbol: clean });
    const raw = await this.transport.request<any>(
      `/api/v1/options/gex?${params.toString()}`,
      "GET",
      undefined,
      options
    );
    const payload = raw?.data || raw;
    const rows = Array.isArray(payload) ? payload : payload?.items || payload?.data;
    const callGex = Array.isArray(rows)
      ? rows.reduce((sum: number, row: any) => sum + Number(row?.call_gex || 0), 0)
      : undefined;
    const putGex = Array.isArray(rows)
      ? rows.reduce((sum: number, row: any) => sum + Number(row?.put_gex || 0), 0)
      : undefined;
    return {
      ...(Array.isArray(payload) ? {} : payload),
      symbol: Array.isArray(payload) ? clean : payload?.symbol || clean,
      net_gex: Array.isArray(payload) ? (callGex || 0) + (putGex || 0) : payload?.net_gex ?? (callGex || 0) + (putGex || 0),
      total_call_gex: Array.isArray(payload) ? callGex : payload?.total_call_gex,
      total_put_gex: Array.isArray(payload) ? putGex : payload?.total_put_gex,
      data: Array.isArray(rows) ? rows : payload?.data,
      items: Array.isArray(rows) ? rows : payload?.items,
    } as OptionGexResponse;
  }

  /**
   * Retrieves overall options market activity and put/call sentiment summary.
   */
  public async getSummary(options?: RequestOptions): Promise<OptionSummaryResponse> {
    const raw = await this.transport.request<any>(
      "/api/v1/options/summary",
      "GET",
      undefined,
      options
    );
    const payload = raw?.data && typeof raw.data === "object" ? raw.data : raw;
    const snapshots = Array.isArray(payload) ? payload : payload?.items || payload?.data || [];
    const first = Array.isArray(snapshots) ? snapshots[0] : snapshots;
    return first && typeof first === "object" ? { ...first, data: snapshots } : { data: snapshots } as OptionSummaryResponse;
  }
}
